File: leetcode/双指针/0142_Linked_List_Cycle_2/Solution.py
class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        fast = slow = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

            if (fast == slow):
                break
        
        if not fast or not fast.next:
            return None
        
        fast = head
        while fast != slow:
            fast, slow = fast.next, slow.next

        return fast


class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

File: leetcode/双指针/0142_Linked_List_Cycle_2/Solution.java
public class Solution {
    
    public ListNode detectCycle(ListNode head) {
        ListNode fast = head, slow = head;

        // 注意! 不能再 while 里判断是否 fast != slow
        // 因为开始时, fast == slow == head
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;

            if (fast == slow) {
                break;
            }
        }

        if (fast == null || fast.next == null) {
            return null;
        }

        fast = head;
        while (fast != slow) {
            fast = fast.next;
            slow = slow.next;
        }

        return fast;
    }

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
            this.next = null;
        }
    }
}
